ethash/src/compute.rs

```diff
@@ -332,21 +358,38 @@ fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64)
 		unroll! {
 			for i in 0..8 {
 				let w = i * 4;
-				let mut reduction = *mix.get_unchecked(0).as_words().get_unchecked(w + 0);
-				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix.get_unchecked(0).as_words().get_unchecked(w + 1);
-				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix.get_unchecked(0).as_words().get_unchecked(w + 2);
-				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix.get_unchecked(0).as_words().get_unchecked(w + 3);
-				*mix.get_unchecked_mut(0).as_words_mut().get_unchecked_mut(i) = reduction;
+				let mut reduction = *mix_words.get_unchecked(w + 0);
+				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix_words.get_unchecked(w + 1);
+				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix_words.get_unchecked(w + 2);
+				reduction = reduction.wrapping_mul(FNV_PRIME) ^ *mix_words.get_unchecked(w + 3);
+				*mix_words.get_unchecked_mut(i) = reduction;
 			}
 		}
 
-		let mut mix_hash = [0u8; 32];
-		let mut buf = [0u8; 32 + 64];
-		ptr::copy_nonoverlapping(f_mix.get_unchecked_mut(0).bytes.as_ptr(), buf.as_mut_ptr(), 64);
-		ptr::copy_nonoverlapping(mix.get_unchecked_mut(0).bytes.as_ptr(), buf[64..].as_mut_ptr(), 32);
-		ptr::copy_nonoverlapping(mix.get_unchecked_mut(0).bytes.as_ptr(), mix_hash.as_mut_ptr(), 32);
-		let mut value: H256 = [0u8; 32];
+		let mix: &mut [Node; MIX_NODES] = make_const_array!(MIX_NODES, mix_words);
+
+		// Explicit use of uninit so we don't have to rely on the optimizer
+		let mut mix_hash: [u8; 32] = mem::uninitialized();
+		let mut buf: [u8; 32 + 64] = mem::uninitialized();
+		ptr::copy_nonoverlapping(
+			f_mix.get_unchecked_mut(0).bytes.as_ptr(),
+			buf.as_mut_ptr(),
+			64,
+		);
+		ptr::copy_nonoverlapping(
+			mix.get_unchecked_mut(0).bytes.as_ptr(),
+			buf[64..].as_mut_ptr(),
+			32,
+		);
+		ptr::copy_nonoverlapping(
+			mix.get_unchecked_mut(0).bytes.as_ptr(),
+			mix_hash.as_mut_ptr(),
+			32,
+		);
```

Comment:
A lot of our `ptr::copy_nonoverlapping` calls can be replaced with `copy_from_slice`, I'll rewrite all our code to use it in a seperate PR (unless you'd consider this a blocker).